boskos/client/client.go

```diff
@@ -640,3 +693,34 @@ func isDialErrorRetriable(err error) bool {
 	}
 	return false
 }
+
+// workFunc describes retrieable work. It should
+// * Return an error for non-recoverable errors
+// * Write retriable errors into `retriedErrs` and return with false, nil
+// * Return with true, nil on success
+type workFunc func(retriedErrs *[]error) (bool, error)
+
+// SleepFunc is called when requests are retried. This may be replaced in tests.
+var SleepFunc = time.Sleep
+
+func retry(work workFunc) error {
+	var retriedErrs []error
+
+	maxAttempts := 4
+	for i := 0; i < maxAttempts; i++ {
+		success, err := work(&retriedErrs)
+		if err != nil {
+			return err
```

Comment:
May be useful to return `append(retriedErrs, err)` to differentiate immediate from eventual failures.

Reply:
The problem is that some of the errors returned by the work funcs return specific errors other places in the code check on. So we would have to first adjust all dependent code to use errors.Is, rather than equal checks

Reply:
(On top of that, utilerrors.Aggregate does not support matching via errors.Is today. I'll make an upstream PR for that so it gets in asap due to their long release cycles)

Reply:
Sounds reasonable.